### ISSUE TITLE
Is it possible to cancel a signal?

### ISSUE THREAD

**Issue by nikek** [cerebral/cerebral#232]
I'm very new to cerebral but I'm very exited about the idea of signals declaratively controlling the flow from intent -> state change and all the mess that goes into that.

But the examples I've seen so far don't present any way of canceling a signal. So just wanted to check if that is possible.

Something like an exposed signal method that drops the current async actions and let's you restart it without waiting for success/fail. Or the reactive takeUntil(thisSignalCalledAgain). Like you might want to do in a typeahead for example.

**Comment by christianalfoni** [comment]
Hi @nikek and thanks for taking a look into Cerebral :-)

Currently there is no way to cancel a signal. It needs to run from start to finish. Its async behaviour is based on promises and those can not be cancelled.

As Observables are taking a stronger foothold these days and patterns for not only transforming values, but actually holding and changing state, we are looking into that for the next version of Cerebral.

So yeah, you get a really nice declarative way of controlling state changes and "the flow of things", but cancellable is not in there. In the example of typeahead you would solve it like you would do with promises. For example creating a reference to the latest request and only set that as the response. For example:

```js
[
  setNewRequestReference,
  getResult, {
    success: [setResultByLastReference],
    error: []
  }
]
```
So basically you can fire off this signal as much as you want, but it will only set the result of the last fired signal. Actually this is a good idea to put into the `http` module. It still will not cancel, but you will at least avoid setting an older response on a newer request.

I hope this helps :-)

**Comment by nikek** [comment]
@christianalfoni Thank you for the quick response! I understand now how it works and how I can solve it the cerebral way :). And yeah, the reactive paradigm is becoming more prominent every day. It could be a good fit for Cerebral, but I know too little to fit it into the behaviour of signals.

I've lately worked with AngularJS and they've got this way of passing a "cancel/timeout promise" to the $http(). But you have this extra deferred.promise to keep track of and connect to some intention of cancelling. It's an ok solution. It makes it possible to abort a request and angular exposes this through their timeout property on the config object in $http.
> timeout – {number|Promise} – timeout in milliseconds, or promise that should abort the request when resolved. (from [$http docs](https://docs.angularjs.org/api/ng/service/$http#usage))

Found that cerebral-http module is using axios and they also have a timeout, but not the promise one.
>   // `timeout` specifies the number of milliseconds before the request times out.
  // If the request takes longer than `timeout`, the request will be aborted.
  timeout: 1000,

Needed to read up on this a bit so found a bit about the XHR API in browsers, it's low level :)
([mdn xhr abort](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest#abort()))

I'll close this ticket because you quickly and nicely answered my question. But in regards of cancelling XHR with abort, I think it's a great feature to have. Especially for apps that are relying a lot on requests to a backend. Eg. typeahead/search or when quickly moving between views that all inits by sending a bunch of heavy requests.

As I mentioned I'm just getting started with cerebral and will gladly continue to try it out with the solution you provided. In general I think you've found a way of handling actions with great potential and I'm happy to give you feedback as a newcomer to the framework.
Thanks!

**Comment by christianalfoni** [comment]
Glad it was helpful :)

You should join us on the discours server. Just go to www.cerebraljs.com and click the chat icon. I am sure other users also needs a solution for this scenario. And thanks for the tips on axios and Angular. I am sure we can implement a `services.http.cancel(url)` type of thing :)
